File: src/agent/protocol.ts
/** Wire protocol shared by the local agent and the browser client.
 *
 *  The browser cannot spawn processes, so every filesystem and git operation is
 *  an RPC to `enc-tool agent`, a second mode of this same binary running on the
 *  user's machine. Transport is a WebSocket on the loopback interface; loopback
 *  is a "potentially trustworthy" origin, so an https:// page may talk to
 *  ws://127.0.0.1 without tripping mixed-content blocking (Chromium + Firefox;
 *  WebKit refuses, hence the capability badge in the UI).
 *
 *  Frames:
 *    client -> agent   Req
 *    agent  -> client  Chunk*  Res          (one Res per Req, always last)
 *    agent  -> client  Push                 (unsolicited; no id)
 *
 *  All paths crossing the wire are POSIX-style and relative to the workspace
 *  root. The agent rejects absolute paths and any `..` segment (see jail.ts).
 */

export interface Req {
  id: number;
  op: string;
  [param: string]: unknown;
}

export type Res =
  | { id: number; ok: true; data: unknown }
  | { id: number; ok: false; error: string; code?: string };

/** Partial result of a long-running op (search hits, transfer progress). */
export interface Chunk {
  id: number;
  chunk: unknown;
}

/** Server-initiated message (filesystem watcher). */
export interface Push {
  event: string;
  data: unknown;
}

export type ServerFrame = Res | Chunk | Push;

// ── op payloads ───────────────────────────────────────────────────────────

export interface AgentInfo {
  agent: "enc-tool";
  version: string;
  platform: string;
  root: string;
  /** Repository root relative to `root`, or null when the folder is not a repo. */
  repo: string | null;
  gitVersion: string | null;
  /** ripgrep is optional — without it search falls back to `git ls-files` + JS scan. */
  ripgrep: string | null;
  /** Recursive fs.watch is unavailable on some Linux kernels; UI hides live updates. */
  watch: boolean;
}

export interface DirEntry {
  name: string;
  dir: boolean;
  /** Present for files only. */
  size?: number;
  mtime?: number;
  /** True for symlinks; `dir` then reflects the link target. */
  link?: boolean;
}

export interface FileRead {
  /** utf-8 text, or null when the file is binary or over the size cap. */
  text: string | null;
  size: number;
  mtime: number;
  binary: boolean;
  tooLarge: boolean;
  /** Byte offset `text` starts at. 0 for an ordinary whole-file read.
   *
   *  A read with `length` is a window onto a file too big to open whole: the
   *  editor shows that slice read-only rather than the "too large" placeholder
   *  that used to be the only answer. The offset is where the agent actually
   *  started, which is not always where it was asked to — a window must not
   *  begin in the middle of a UTF-8 character. */
  offset: number;
  /** Whether `text` runs to the end of the file. */
  eof: boolean;
}

/** One line of `git reflog`: where a ref has been, and how it got there. */
export interface ReflogEntry {
  /** "HEAD@{3}" — what to pass back to reset onto this point. */
  selector: string;
  oid: string;
  /** The verb git recorded: "commit", "reset", "checkout", "merge"… */
  action: string;
  /** The rest of the line: a commit subject, "moving to main", and so on. */
  message: string;
  /** Unix seconds. */
  time: number;
}

/** One row of `git status --porcelain=v2`. */
export interface StatusEntry {
  path: string;
  /** Previous path for renames/copies. */
  from?: string;
  /** Index (staged) state: one of ".MADRCU" */
  index: string;
  /** Worktree (unstaged) state: one of ".MADRCU" */
  work: string;
  untracked?: boolean;
  ignored?: boolean;
  conflict?: boolean;
}

/** A git operation that stopped part-way and is waiting to be finished.
 *
 *  Merge, rebase, cherry-pick and revert all leave the repository in a state
 *  that only `--continue` or `--abort` gets out of, and during a rebase HEAD is
 *  detached — so without this the UI can only report "(detached)" and leave the
 *  user to work out what happened to their branch.
 */
export interface GitOperation {
  kind: "merge" | "rebase" | "cherry-pick" | "revert";
  /** Branch the operation started on: the one a rebase will restore at the end.
   *  Absent for a merge, where HEAD stays on the branch throughout. */
  head?: string;
  /** What the work is being replayed onto (rebase), or the subject line git
   *  prepared (merge, cherry-pick, revert) — whichever names the other side. */
  onto?: string;
  /** Position in a multi-commit rebase: commit `step` of `total`. */
  step?: number;
  total?: number;
}

export interface GitStatus {
  branch: string | null;
  /** null for a detached HEAD or an unborn branch. */
  upstream: string | null;
  ahead: number;
  behind: number;
  oid: string | null;
  entries: StatusEntry[];
  /** Set while a merge/rebase/cherry-pick/revert is half-finished. */
  operation?: GitOperation | null;
  /** The commit message git has already written for the next commit, with its
   *  comment lines stripped: `MERGE_MSG` after a merge or a cherry-pick that
   *  stopped, `SQUASH_MSG` after `merge --squash`. Null when there is none.
   *
   *  A squash leaves no half-finished operation — the index is staged and git is
   *  done — so this cannot live on `operation`: it is a property of what the
   *  next commit is going to be, not of an operation in progress. */
  preparedMessage?: string | null;
}

export interface Commit {
  oid: string;
  parents: string[];
  author: string;
  email: string;
  /** Author time, seconds since epoch. */
  time: number;
  /** Decoration from %D, e.g. "HEAD -> main, origin/main, tag: v2". */
  refs: string;
  subject: string;
}

export interface Branch {
  /** Full refname, e.g. "refs/heads/main" or "refs/remotes/origin/main". */
  ref: string;
  /** Short display name, e.g. "main" or "origin/main". */
  name: string;
  oid: string;
  upstream: string | null;
  remote: boolean;
  head: boolean;
  /** refs/tags/… — a fixed point rather than something you commit onto. Listed
   *  alongside branches because everything that takes a ref (compare, create a
   *  branch from, check out) takes a tag just as happily, and a picker that
   *  hides them sends people to the terminal for a release. */
  tag: boolean;
  /** Commits this branch has that its upstream does not, and the other way
   *  round. Both 0 when there is no upstream to compare against. */
  ahead: number;
  behind: number;
  /** When the ref last moved, in unix seconds — "3 days ago" beside a branch
   *  name is most of what tells stale from current in a long list. */
  time: number;
}

/** A pair of texts for @codemirror/merge. `before`/`after` are null when the
 *  file does not exist on that side (added / deleted). */
export interface DiffPair {
  path: string;
  before: string | null;
  after: string | null;
  beforeLabel: string;
  afterLabel: string;
  binary: boolean;
}

export interface CommitFile {
  path: string;
  from?: string;
  status: "A" | "M" | "D" | "R" | "C" | "T";
  added: number;
  deleted: number;
  binary: boolean;
}

export interface CommitDetail {
  commit: Commit;
  body: string;
  files: CommitFile[];
  /** For a merge: which parent the file list was diffed against, 1-based.
   *  Absent for an ordinary commit, which has only one side to compare with. */
  parent?: number;
}

/** One line of `git blame`. The browser mirror of this is `BlameLine` in
 *  code/blame.ts, which is where the gutter renders it. */
export interface BlameRow {
  oid: string;
  author: string;
  /** Author time, seconds since epoch. */
  time: number;
  /** 1-based line in the committed file. */
  line: number;
  /** Subject of the commit this line came from, as git's porcelain reports it. */
  summary: string;
}

export interface SearchHit {
  path: string;
  /** 1-based. */
  line: number;
  /** 0-based byte column of the first match on the line. */
  col: number;
  text: string;
  /** [start, end) byte offsets of every match within `text`. */
  ranges: [number, number][];
}

export interface SearchSummary {
  files: number;
  matches: number;
  /** True when the scan stopped at the result cap. */
  truncated: boolean;
  engine: "ripgrep" | "fallback";
}

export interface FsChange {
  /** Deduplicated, workspace-relative paths that changed since the last push. */
  paths: string[];
}

File: agent-go/protocol.go
// Wire protocol shared with the browser client.
//
// The authoritative definition lives in src/agent/protocol.ts — the browser
// side is written against those types, so these structs exist to match them
// byte for byte on the wire. Every json tag here is load-bearing: a renamed
// field is a silently empty panel, not a compile error. scripts/protocol-check
// diffs the two files' field lists on every build so a drift cannot ship.
//
// Nullable fields are pointers on purpose. `size` on an empty file is 0 and
// must still be sent, while `size` on a directory must be absent — omitempty on
// a plain int cannot express that difference.
package main

import "encoding/json"

// ── frames ────────────────────────────────────────────────────────────────

type req struct {
	id     int64
	op     string
	params map[string]json.RawMessage
}

type resOK struct {
	ID   int64 `json:"id"`
	OK   bool  `json:"ok"`
	Data any   `json:"data"`
}

type resErr struct {
	ID    int64  `json:"id"`
	OK    bool   `json:"ok"`
	Error string `json:"error"`
	Code  string `json:"code,omitempty"`
}

// Partial result of a long-running op (search hits, transfer progress).
type chunkFrame struct {
	ID    int64 `json:"id"`
	Chunk any   `json:"chunk"`
}

// Server-initiated message (filesystem watcher).
type pushFrame struct {
	Event string `json:"event"`
	Data  any    `json:"data"`
}

// ── op payloads ───────────────────────────────────────────────────────────

type agentInfo struct {
	Agent    string `json:"agent"`
	Version  string `json:"version"`
	Platform string `json:"platform"`
	Root     string `json:"root"`
	// Repository root relative to Root, or null when the folder is not a repo.
	Repo       *string `json:"repo"`
	GitVersion *string `json:"gitVersion"`
	// ripgrep is optional — without it search falls back to `git ls-files`.
	Ripgrep *string `json:"ripgrep"`
	// Recursive watching is unavailable on some hosts; UI hides live updates.
	Watch bool `json:"watch"`
}

type dirEntry struct {
	Name string `json:"name"`
	Dir  bool   `json:"dir"`
	// Present for files only.
	Size  *int64 `json:"size,omitempty"`
	Mtime *int64 `json:"mtime,omitempty"`
	// True for symlinks; Dir then reflects the link target.
	Link bool `json:"link,omitempty"`
}

type fileRead struct {
	// utf-8 text, or null when the file is binary or over the size cap.
	Text     *string `json:"text"`
	Size     int64   `json:"size"`
	Mtime    int64   `json:"mtime"`
	Binary   bool    `json:"binary"`
	TooLarge bool    `json:"tooLarge"`
}

// One row of `git status --porcelain=v2`.
type statusEntry struct {
	Path string `json:"path"`
	// Previous path for renames/copies.
	From string `json:"from,omitempty"`
	// Index (staged) state: one of ".MADRCU"
	Index string `json:"index"`
	// Worktree (unstaged) state: one of ".MADRCU"
	Work      string `json:"work"`
	Untracked bool   `json:"untracked,omitempty"`
	Ignored   bool   `json:"ignored,omitempty"`
	Conflict  bool   `json:"conflict,omitempty"`
}

type gitStatus struct {
	Branch *string `json:"branch"`
	// null for a detached HEAD or an unborn branch.
	Upstream *string       `json:"upstream"`
	Ahead    int           `json:"ahead"`
	Behind   int           `json:"behind"`
	Oid      *string       `json:"oid"`
	Entries  []statusEntry `json:"entries"`
}

type commit struct {
	Oid     string   `json:"oid"`
	Parents []string `json:"parents"`
	Author  string   `json:"author"`
	Email   string   `json:"email"`
	// Author time, seconds since epoch.
	Time int64 `json:"time"`
	// Decoration from %D, e.g. "HEAD -> main, origin/main, tag: v2".
	Refs    string `json:"refs"`
	Subject string `json:"subject"`
}

type branch struct {
	// Full refname, e.g. "refs/heads/main" or "refs/remotes/origin/main".
	Ref string `json:"ref"`
	// Short display name, e.g. "main" or "origin/main".
	Name     string  `json:"name"`
	Oid      string  `json:"oid"`
	Upstream *string `json:"upstream"`
	Remote   bool    `json:"remote"`
	Head     bool    `json:"head"`
}

// A pair of texts for @codemirror/merge. Before/After are null when the file
// does not exist on that side (added / deleted).
type diffPair struct {
	Path        string  `json:"path"`
	Before      *string `json:"before"`
	After       *string `json:"after"`
	BeforeLabel string  `json:"beforeLabel"`
	AfterLabel  string  `json:"afterLabel"`
	Binary      bool    `json:"binary"`
}

type commitFile struct {
	Path    string `json:"path"`
	From    string `json:"from,omitempty"`
	Status  string `json:"status"`
	Added   int    `json:"added"`
	Deleted int    `json:"deleted"`
	Binary  bool   `json:"binary"`
}

type commitDetail struct {
	Commit commit       `json:"commit"`
	Body   string       `json:"body"`
	Files  []commitFile `json:"files"`
}

type blameRow struct {
	Oid    string `json:"oid"`
	Author string `json:"author"`
	Time   int64  `json:"time"`
	Line   int    `json:"line"`
}

type searchHit struct {
	Path string `json:"path"`
	// 1-based.
	Line int `json:"line"`
	// 0-based column of the first match on the line.
	Col  int    `json:"col"`
	Text string `json:"text"`
	// [start, end) offsets of every match within Text.
	Ranges [][2]int `json:"ranges"`
}

type searchSummary struct {
	Files   int `json:"files"`
	Matches int `json:"matches"`
	// True when the scan stopped at the result cap.
	Truncated bool `json:"truncated"`
	// "ripgrep" or "fallback".
	Engine string `json:"engine"`
}
